File: stone_processor/src/main/java/com/github/klee0kai/stone/helpers/invokecall/ModulesGraph.java
package com.github.klee0kai.stone.helpers.invokecall;

import com.github.klee0kai.stone.AnnotationProcessor;
import com.github.klee0kai.stone._hidden_.provide.ProvideBuilder;
import com.github.klee0kai.stone._hidden_.types.CacheAction;
import com.github.klee0kai.stone._hidden_.types.ListUtils;
import com.github.klee0kai.stone.exceptions.IncorrectSignatureException;
import com.github.klee0kai.stone.exceptions.ObjectNotProvidedException;
import com.github.klee0kai.stone.exceptions.RecursiveProviding;
import com.github.klee0kai.stone.exceptions.StoneException;
import com.github.klee0kai.stone.model.ClassDetail;
import com.github.klee0kai.stone.model.FieldDetail;
import com.github.klee0kai.stone.model.MethodDetail;
import com.github.klee0kai.stone.model.annotations.BindInstanceAnn;
import com.github.klee0kai.stone.model.annotations.ProvideAnn;
import com.github.klee0kai.stone.model.annotations.QualifierAnn;
import com.github.klee0kai.stone.utils.RecursiveDetector;
import com.github.klee0kai.stone.wrappers.Ref;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;

import java.util.*;

import static com.github.klee0kai.stone.AnnotationProcessor.allClassesHelper;
import static com.github.klee0kai.stone.codegen.ModuleCacheControlInterfaceBuilder.cacheControlMethodName;
import static com.github.klee0kai.stone.exceptions.ExceptionStringBuilder.createErrorMes;
import static com.github.klee0kai.stone.helpers.invokecall.InvokeCall.INVOKE_PROVIDE_BIND_INSTANCE;
import static com.github.klee0kai.stone.helpers.invokecall.InvokeCall.INVOKE_PROVIDE_OBJECT_CACHED;
import static com.github.klee0kai.stone.helpers.wrap.WrapHelper.*;
import static com.github.klee0kai.stone.utils.LocalFieldName.genLocalFieldName;

public class ModulesGraph {

    public static boolean SIMPLE_PROVIDE_OPTIMIZING = true;
    public static int MAX_PROVIDE_RESOLVE_COUNT = 10_000;

    private final HashMap<TypeName, Set<InvokeCall>> provideTypeCodes = new HashMap<>();
    private final HashMap<TypeName, Set<InvokeCall>> cacheControlTypeCodes = new HashMap<>();


    /**
     * Methods graph build.
     *
     * @param provideModuleMethod module's provide method
     * @param module              module's class
     */
    public void collectFromModule(MethodDetail provideModuleMethod, ClassDetail module) {
        ClassDetail iModuleInterface = AnnotationProcessor.allClassesHelper.iModule;
        for (MethodDetail m : module.getAllMethods(false, true, "<init>")) {
            TypeName provTypeName = nonWrappedType(m.returnType);
            if (provTypeName.isPrimitive() || provTypeName == TypeName.VOID)
                continue;
            if (iModuleInterface.findMethod(m, false) != null)
                continue;
            boolean isCached = !m.hasAnnotations(ProvideAnn.class) || m.ann(ProvideAnn.class).isCachingProvideType();
            boolean isBindInstance = m.hasAnnotations(BindInstanceAnn.class);
            int invokeProvideFlags = isCached ? INVOKE_PROVIDE_OBJECT_CACHED : 0;
            invokeProvideFlags |= isBindInstance ? INVOKE_PROVIDE_BIND_INSTANCE : 0;

            provideTypeCodes.putIfAbsent(provTypeName, new HashSet<>());
            provideTypeCodes.get(provTypeName).add(new InvokeCall(invokeProvideFlags, provideModuleMethod, m));

            MethodDetail cacheControlMethod = new MethodDetail();
            cacheControlMethod.methodName = cacheControlMethodName(m.methodName);
            cacheControlMethod.args.add(FieldDetail.simple("__action", ClassName.get(CacheAction.class)));
            for (FieldDetail it : m.args) {
                if (!((it.type instanceof ClassName) && allClassesHelper.allIdentifiers.contains(it.type)))
                    continue;
                cacheControlMethod.args.add(it);
            }
            cacheControlMethod.returnType = listWrapTypeIfNeed(m.returnType);
            cacheControlMethod.qualifierAnns = m.qualifierAnns;

            cacheControlTypeCodes.putIfAbsent(provTypeName, new HashSet<>());
            cacheControlTypeCodes.get(provTypeName).add(new InvokeCall(provideModuleMethod, cacheControlMethod));
        }
    }

    public CodeBlock codeProvideType(
            String methodName,
            TypeName returnType,
            Set<QualifierAnn> qualifierAnns,
            Collection<FieldDetail> declaredFields
    ) {
        boolean isWrappedReturn = isSupport(returnType);
        TypeName providingType = isWrappedReturn ? nonWrappedType(returnType) : returnType;

        Set<ProvideDep> provideDeps = new HashSet<>();
        provideDeps.add(new ProvideDep(methodName, returnType, qualifierAnns));
        List<InvokeCall> provideTypeInvokes = provideInvokesWithDeps(provideDeps.iterator().next());
        if (provideTypeInvokes == null || provideTypeInvokes.isEmpty()) {
            return null;
        }
        for (InvokeCall provideTypeInvoke : provideTypeInvokes) provideDeps.addAll(provideTypeInvoke.argDeps());
        if (SIMPLE_PROVIDE_OPTIMIZING && provideTypeInvokes.size() == 1 && !isList(returnType)) {
            InvokeCall invokeCall = provideTypeInvokes.get(0).best();
            return transform(
                    invokeCall.rawReturnType(),
                    returnType,
                    invokeCall.invokeCode(declaredFields)
            );
        }

        TypeName provideBuilder = ParameterizedTypeName.get(ClassName.get(ProvideBuilder.class), providingType);
        TypeName provideBuilderList = ParameterizedTypeName.get(ClassName.get(Collection.class), providingType);
        String listFieldName = genLocalFieldName();
        List<FieldDetail> localVariables = new LinkedList<>(declaredFields);

        CodeBlock.Builder codeBlock = CodeBlock.builder();
        codeBlock.add("new $T( ( $L ) -> { \n", provideBuilder, listFieldName);

        for (InvokeCall inv : provideTypeInvokes) {
            boolean isCacheProvide = (inv.flags & INVOKE_PROVIDE_OBJECT_CACHED) != 0;
            boolean isSingleDepRequired = ListUtils.contains(provideDeps, (idx, it) ->
                    Objects.equals(nonWrappedType(it.typeName), nonWrappedType(inv.resultType()))
                            && !isList(it.typeName)
            );
            boolean isListDepRequired = ListUtils.contains(provideDeps, (idx, it) ->
                    Objects.equals(nonWrappedType(it.typeName), nonWrappedType(inv.resultType()))
                            && isList(it.typeName)
            );
            FieldDetail singleDepField = FieldDetail.simple(genLocalFieldName(), inv.resultType());
            singleDepField.qualifierAnns = inv.qualifierAnnotations(true);

            FieldDetail listDepField = FieldDetail.simple(
                    genLocalFieldName(),
                    ParameterizedTypeName.get(ClassName.get(Ref.class), ParameterizedTypeName.get(ClassName.get(List.class), inv.resultType())));
            listDepField.qualifierAnns = inv.qualifierAnnotations(true);


            if (isSingleDepRequired) {
                if (isCacheProvide) {
                    codeBlock.add("$T $L = ", inv.resultType(), singleDepField.name)
                            .addStatement(
                                    transform(
                                            inv.best().rawReturnType(),
                                            inv.resultType(),
                                            inv.best().invokeCode(localVariables))
                            );


                    localVariables.add(singleDepField);
                } else {
                    singleDepField.type = ParameterizedTypeName.get(ClassName.get(Ref.class), inv.resultType());
                    codeBlock.add("$T $L = () -> ", singleDepField.type, singleDepField.name)
                            .addStatement(transform(inv.best().rawReturnType(), inv.resultType(), inv.best().invokeCode(localVariables)));

                    localVariables.add(singleDepField);
                }
            }

            if (isListDepRequired) {
                codeBlock.add("$T $L = () -> ", listDepField.type, listDepField.name)
                        .addStatement(inv.invokeAllToList(localVariables));
                localVariables.add(listDepField);
            }


            if (Objects.equals(inv.resultType(), providingType)) {
                if (isList(returnType)) {
                    codeBlock.add("$L.addAll( $L );\n", listFieldName,
                            transform(
                                    listDepField.type,
                                    provideBuilderList,
                                    CodeBlock.of(listDepField.name)
                            )
                    );
                } else {
                    codeBlock.add("$L.add( $L );\n", listFieldName,
                            transform(
                                    singleDepField.type,
                                    providingType,
                                    CodeBlock.of(singleDepField.name)
                            )
                    );
                }
                if (!isList(returnType))
                    break;
            }
        }

        codeBlock.add("\n  })");
        if (isList(returnType)) {
            codeBlock.add(".all() ");

            return transform(
                    ParameterizedTypeName.get(ClassName.get(List.class), providingType),
                    returnType,
                    codeBlock.build()
            );
        } else {
            codeBlock.add(".first() ");

            return transform(
                    providingType,
                    returnType,
                    codeBlock.build()
            );
        }

    }


    public List<InvokeCall> provideInvokesWithDeps(ProvideDep provideDep) {
        LinkedList<InvokeCall> provideTypeInvokes = new LinkedList<>();
        LinkedList<ProvideDep> needProvideDeps = new LinkedList<>();
        RecursiveDetector<Integer> needProvideDepsRecursiveDetector = new RecursiveDetector<>();
        needProvideDeps.add(provideDep);
        int loopCount = 0;

        // provide dependencies while not provide all
        while (!needProvideDeps.isEmpty()) {
            ProvideDep rawDep = needProvideDeps.pollFirst();
            TypeName dep = nonWrappedType(rawDep.typeName);
            InvokeCall invokeCall = provideTypeInvokeCall(provideTypeCodes, dep, rawDep.qualifierAnns, rawDep.methodName, isList(rawDep.typeName));
            if (invokeCall == null) {
                if (Objects.equals(provideDep, rawDep)) {
                    return null;
                }

                throw new ObjectNotProvidedException(
                        createErrorMes()
                                .errorProvideType(dep.toString())
                                .build()
                );
            }

            boolean isBindInstanceInvoke = (invokeCall.flags & INVOKE_PROVIDE_BIND_INSTANCE) != 0;
            List<ProvideDep> newDeps = ListUtils.filter(invokeCall.argDeps(), (i, it) -> {
                if (isBindInstanceInvoke && Objects.equals(rawDep.typeName, it.typeName)) {
                    // bind instance case. Argument and return type are equals
                    return false;
                }
                // qualifies not need to provide
                TypeName argNonWrapped = nonWrappedType(it.typeName);
                return argNonWrapped instanceof ClassName && !allClassesHelper.allIdentifiers.contains(argNonWrapped);
            });

            needProvideDeps.addAll(newDeps);
            needProvideDeps = ListUtils.removeDoublesRight(needProvideDeps, Objects::equals);
            boolean recursiveDetected = !newDeps.isEmpty() && needProvideDepsRecursiveDetector.next(needProvideDeps.hashCode());
            if (recursiveDetected) {
                throw new RecursiveProviding(
                        createErrorMes()
                                .errorProvideType(provideDep.typeName.toString())
                                .recursiveProviding()
                                .build()
                );
            }
            if (loopCount++ > MAX_PROVIDE_RESOLVE_COUNT) {
                throw new StoneException(
                        createErrorMes()
                                .errorProvideType(provideDep.typeName.toString())
                                .add("long providing loop for type. Stone library Error.")
                                .build(),
                        null
                );
            }

            provideTypeInvokes.add(invokeCall);
            provideTypeInvokes = ListUtils.removeDoublesRight(provideTypeInvokes, (it1, it2) -> {
                return Objects.equals(it1.resultType(), it2.resultType())
                        && Objects.equals(it1.qualifierAnnotations(true), it2.qualifierAnnotations(true));
            });

        }
        Collections.reverse(provideTypeInvokes);
        return provideTypeInvokes;
    }

    /**
     * Generate cache control method invoke. Clean refs, change ref type and other
     *
     * @param provideMethodName predefined method name
     * @param typeName          the name of the type whose cache needs to be changed
     * @return cache control invoke call
     */
    public InvokeCall invokeControlCacheForType(String provideMethodName, TypeName typeName, Set<QualifierAnn> qualifierAnns) {
        String cacheControlMethodName = cacheControlMethodName(provideMethodName);
        return provideTypeInvokeCall(cacheControlTypeCodes, typeName, qualifierAnns, cacheControlMethodName, false);
    }

    private InvokeCall provideTypeInvokeCall(
            HashMap<TypeName, Set<InvokeCall>> provideTypeCodes,
            TypeName typeName,
            Set<QualifierAnn> qualifierAnns,
            String provideMethodName,
            boolean listVariants
    ) {
        Set<InvokeCall> invokeCalls = provideTypeCodes.getOrDefault(typeName, null);
        if (invokeCalls == null || invokeCalls.isEmpty())
            return null;
        List<InvokeCall> filtered = !listVariants || qualifierAnns != null && !qualifierAnns.isEmpty()
                ? ListUtils.filter(invokeCalls, (i, it) -> Objects.equals(it.qualifierAnnotations(false), qualifierAnns))
                : new LinkedList<>(invokeCalls);

        filtered = provideMethodName != null ? ListUtils.filter(filtered, (i, it) -> {
            int len = it.bestSequence().size();
            String mName = it.bestSequence().get(len - 1).methodName;
            return Objects.equals(provideMethodName, mName);
        }) : filtered;

        if (!listVariants && filtered.size() > 1) {
            throw new IncorrectSignatureException(
                    createErrorMes()
                            .errorProvideType(typeName.toString())
                            .add(": is bound multi times.\n")
                            .add(String.join(" and \n", ListUtils.format(filtered, InvokeCall::toString)))
                            .build());
        }
        return !filtered.isEmpty() ? new InvokeCall(filtered) : null;
    }

}

File: stone_ksp/src/commonMain/kotlin/com/github/klee0kai/thekey/stone/ksp/helpers/invokecall/ModulesGraph.kt
@file:OptIn(KspExperimental::class)

package com.github.klee0kai.thekey.stone.ksp.helpers.invokecall

import com.github.klee0kai.stone.__hidden__.CacheAction
import com.github.klee0kai.stone.__hidden__.provide.ProvideBuilder
import com.github.klee0kai.stone.annotations.module.BindInstance
import com.github.klee0kai.stone.annotations.module.Provide
import com.github.klee0kai.stone.annotations.qualifier.IgnoreQualifier
import com.github.klee0kai.stone.weakref.Ref
import com.github.klee0kai.thekey.stone.ksp.exceptions.*
import com.github.klee0kai.thekey.stone.ksp.helpers.cacheControlMethodName
import com.github.klee0kai.thekey.stone.ksp.helpers.hiddenModuleStoneClName
import com.github.klee0kai.thekey.stone.ksp.helpers.identifierParameters
import com.github.klee0kai.thekey.stone.ksp.helpers.invokecall.model.*
import com.github.klee0kai.thekey.stone.ksp.helpers.qualifierAnnotations
import com.github.klee0kai.thekey.stone.ksp.helpers.wrap.WrapHelper
import com.github.klee0kai.thekey.stone.ksp.ksp.getAllMethods
import com.github.klee0kai.thekey.stone.ksp.ksp.isNotPrimitive
import com.github.klee0kai.thekey.stone.ksp.ksp.resolveNotNullable
import com.github.klee0kai.thekey.stone.ksp.target.component.*
import com.github.klee0kai.thekey.stone.ksp.utils.LocalFieldName.genLocalFieldName
import com.github.klee0kai.thekey.stone.ksp.utils.RecursiveDetector
import com.github.klee0kai.thekey.stone.ksp.utils.removeDoubles
import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getAnnotationsByType
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import java.util.*

class ModulesGraph(
    val wrapHelper: WrapHelper,
    val identifierTypes: List<KSType>,
) {

    private val provideTypeCodes = HashMap<TypeName, HashSet<InvokeCall>>()
    private val cacheControlTypeCodes = HashMap<TypeName, HashSet<InvokeCall>>()


    fun collectFromComponent(
        componentCl: KSClassDeclaration,
    ) {
        val componentsAllMethods = componentCl
            .getAllMethods(includeObjectMethods = false, allowDoubles = false, "<init>")
        val hiddenModuleProvideMethod = MethodDetail(
            methodName = GenComponentProcessor.hiddenModuleFieldName,
            returnType = componentCl.hiddenModuleStoneClName,
            isProperty = true,
        )

        componentsAllMethods
            .filter { it.isModuleProvideMethod || it.isDepsProvideMethod }
            .forEachFun { _, moduleProvideMethod ->

                val module = moduleProvideMethod.returnType?.resolveNotNullable()
                    ?.declaration as? KSClassDeclaration ?: return@forEachFun

                for (m in module.getAllMethods(includeObjectMethods = false, allowDoubles = true, "<init>")) {
                    if (m.returnType?.resolveNotNullable()?.isNotPrimitive == false) continue

                    val returnType = m.returnType?.resolveNotNullable()?.toTypeName() ?: continue
                    val provTypeName = wrapHelper.nonWrappedType(returnType)
                    val isCached = m.getAnnotationsByType(Provide::class)
                        .firstOrNull()?.cache !in listOf(Provide.CacheType.Factory, null)
                    val isBindInstance = m.getAnnotationsByType(BindInstance::class).firstOrNull() != null

                    provideTypeCodes.putIfAbsent(provTypeName, HashSet<InvokeCall>())
                    provideTypeCodes[provTypeName]?.add(
                        InvokeCall.fromSequence(
                            wrapHelper = wrapHelper,
                            callSequence = listOf(moduleProvideMethod.toMethodDetail(), m.toMethodDetail()),
                            flags = InvokeProvideFlags(
                                provideObjectCached = isCached,
                                provideBindInstance = isBindInstance,
                            ),
                        )
                    )

                    val cacheControlMethod = MethodDetail(
                        methodName = m.cacheControlMethodName,
                        returnType = wrapHelper.listWrapTypeIfNeed(returnType),
                        qualifierAnns = m.qualifierAnnotations.map { it.toQualifierAnn() }.toSet(),
                        args = buildList {
                            add(FieldDetail.simple("__action", CacheAction::class.asClassName()))
                            addAll(m.parameters.identifierParameters(identifierTypes).map { it.toFieldDetail() })
                        },
                    )

                    cacheControlTypeCodes.putIfAbsent(provTypeName, HashSet<InvokeCall>())
                    cacheControlTypeCodes[provTypeName]?.add(
                        InvokeCall.fromSequence(
                            wrapHelper = wrapHelper,
                            callSequence = listOf(moduleProvideMethod.toMethodDetail(), cacheControlMethod)
                        )
                    )
                }
            }


        componentsAllMethods
            .filter { it.isBindInstanceMethod == BindInstanceType.BindInstanceAndProvide }
            .forEachFun { _, componentMethod ->
                val returnType = componentMethod.returnType?.resolve()?.toTypeName() ?: return@forEachFun
                val nonInModules = provideInvokesWithDeps(
                    ProvideDep(
                        componentMethod.simpleName.asString(),
                        wrapHelper.listWrapTypeIfNeed(returnType),
                        componentMethod.qualifierAnnotations.map { it.toQualifierAnn() }.toSet()
                    )
                ) == null
                if (!nonInModules) return@forEachFun

                val provTypeName = wrapHelper.nonWrappedType(returnType)
                val isCached = componentMethod.getAnnotationsByType(Provide::class)
                    .firstOrNull()?.cache !in listOf(Provide.CacheType.Factory, null)
                val isBindInstance = componentMethod.getAnnotationsByType(BindInstance::class).firstOrNull() != null

                provideTypeCodes.putIfAbsent(provTypeName, HashSet<InvokeCall>())
                provideTypeCodes[provTypeName]?.add(
                    InvokeCall.fromSequence(
                        wrapHelper = wrapHelper,
                        callSequence = listOf(hiddenModuleProvideMethod, componentMethod.toMethodDetail()),
                        flags = InvokeProvideFlags(
                            provideObjectCached = isCached,
                            provideBindInstance = isBindInstance,
                        ),
                    )
                )

                val cacheControlMethod = MethodDetail(
                    methodName = componentMethod.cacheControlMethodName,
                    returnType = wrapHelper.listWrapTypeIfNeed(returnType),
                    qualifierAnns = componentMethod.qualifierAnnotations.map { it.toQualifierAnn() }.toSet(),
                    args = buildList {
                        add(FieldDetail.simple("__action", CacheAction::class.asClassName()))
                        addAll(
                            componentMethod.parameters.identifierParameters(identifierTypes)
                                .map { it.toFieldDetail() })
                    },
                )

                cacheControlTypeCodes.putIfAbsent(provTypeName, HashSet<InvokeCall>())
                cacheControlTypeCodes[provTypeName]
                    ?.add(
                        InvokeCall.fromSequence(
                            wrapHelper = wrapHelper,
                            callSequence = listOf(hiddenModuleProvideMethod, cacheControlMethod)
                        )
                    )
            }
    }


    fun codeProvideType(
        methodName: String?,
        returnType: TypeName,
        qualifierAnns: Set<QualifierAnn>,
        declaredFields: List<FieldDetail>
    ): CodeBlock? {
        val isWrappedReturn = wrapHelper.isSupport(returnType)
        val providingType = if (isWrappedReturn) wrapHelper.nonWrappedType(returnType) else returnType
        val provideDeps = HashSet<ProvideDep>()
        provideDeps.add(
            ProvideDep(
                methodName = methodName,
                typeName = wrapHelper.listWrapTypeIfNeed(returnType),
                qualifierAnns = qualifierAnns,
            )
        )
        val provideTypeInvokes = provideInvokesWithDeps(provideDeps.iterator().next())
        if (provideTypeInvokes.isNullOrEmpty()) return null

        for (provideTypeInvoke in provideTypeInvokes) provideDeps.addAll(provideTypeInvoke.argDeps())
        if (provideTypeInvokes.size == 1 && !wrapHelper.isList(returnType)) {
            val invokeCall = provideTypeInvokes.first().best()
            return wrapHelper.transform(
                invokeCall.rawReturnType(),
                returnType,
                invokeCall.invokeCode(declaredFields)
            )
        }

        val provideBuilder = ProvideBuilder::class.asClassName().parameterizedBy(providingType)
        val provideBuilderList = Collection::class.asClassName().parameterizedBy(providingType)

        val listFieldName = genLocalFieldName()
        val localVariables = LinkedList<FieldDetail>(declaredFields)

        val codeBlock = CodeBlock.builder()
        codeBlock.add("%T{ %L -> \n", provideBuilder, listFieldName)

        for (inv in provideTypeInvokes) {
            val isCacheProvide = inv.flags.provideObjectCached
            val isSingleDepRequired = provideDeps.any {
                wrapHelper.nonWrappedType(it.typeName) == wrapHelper.nonWrappedType(inv.resultType())
                        && !wrapHelper.isList(it.typeName)
            }
            val isListDepRequired = provideDeps.any {
                wrapHelper.nonWrappedType(it.typeName) == wrapHelper.nonWrappedType(inv.resultType())
                        && wrapHelper.isList(it.typeName)
            }
            var singleDepField = FieldDetail(
                name = genLocalFieldName(),
                type = inv.resultType(),
                qualifierAnns = inv.qualifierAnnotations
            )
            val listDepField = FieldDetail(
                name = genLocalFieldName(),
                type = Ref::class.asClassName().parameterizedBy(
                    List::class.asClassName().parameterizedBy(
                        inv.resultType()
                    )
                ),
                qualifierAnns = inv.qualifierAnnotations
            )

            if (isSingleDepRequired) {
                if (isCacheProvide) {
                    codeBlock.add("// 1 ${inv.qualifierAnnotations.joinToString(",") { it.logString() }} \n")
                    codeBlock.add("val %L = ", singleDepField.name)
                        .add(
                            wrapHelper.transform(
                                inv.best().rawReturnType(),
                                inv.resultType(),
                                inv.best().invokeCode(localVariables)
                            )
                        )
                        .addStatement("")


                    localVariables.add(singleDepField)
                } else {
                    singleDepField = singleDepField
                        .copy(type = Ref::class.asClassName().parameterizedBy(inv.resultType()))

                    codeBlock.add("// 2 ${inv.qualifierAnnotations.joinToString(",") { it.logString() }} \n")
                    codeBlock.add("val %L = %T{ ", singleDepField.name, Ref::class)
                        .add(
                            wrapHelper.transform(
                                inv.best().rawReturnType(),
                                inv.resultType(),
                                inv.best().invokeCode(localVariables),
                            )
                        )
                        .addStatement(" } ")

                    localVariables.add(singleDepField)
                }
            }

            if (isListDepRequired) {
                codeBlock.add("// 3 ${inv.qualifierAnnotations.joinToString(",") { it.logString() }} \n")
                codeBlock.add("val %L = %T{ ", listDepField.name, listDepField.type)
                    .add(inv.invokeAllToList(localVariables))
                    .addStatement(" } ")

                localVariables.add(listDepField)
            }


            if (inv.resultType().copy(nullable = false) == providingType.copy(nullable = false)) {
                if (wrapHelper.isList(returnType)) {
                    codeBlock.add(
                        "%L.addAll( %L ?: emptyList() )\n",
                        listFieldName,
                        wrapHelper.transform(
                            listDepField.type,
                            provideBuilderList.copy(nullable = true),
                            CodeBlock.of(listDepField.name)
                        )
                    )
                } else {
                    codeBlock.add(
                        "%L.add( %L )\n",
                        listFieldName,
                        wrapHelper.transform(
                            singleDepField.type,
                            providingType.copy(nullable = true),
                            CodeBlock.of(singleDepField.name)
                        )
                    )
                }
                if (!wrapHelper.isList(returnType)) break
            }
        }

        codeBlock.add(" }")
        if (wrapHelper.isList(returnType)) {
            codeBlock.add(".all() ")

            return wrapHelper.transform(
                List::class.asClassName().parameterizedBy(providingType),
                returnType,
                codeBlock.build()
            )

        } else {
            codeBlock.add(if (providingType.isNullable) ".firstOrNull()" else ".first() ")

            return wrapHelper.transform(
                providingType,
                returnType,
                codeBlock.build()
            )
        }
    }


    fun provideInvokesWithDeps(provideDep: ProvideDep): List<InvokeCall>? {
        var provideTypeInvokes = LinkedList<InvokeCall>()
        var needProvideDeps = LinkedList<ProvideDep>()
        val needProvideDepsRecursiveDetector = RecursiveDetector<Int>()
        needProvideDeps.add(provideDep)
        var loopCount = 0

        // provide dependencies while not provide all
        while (!needProvideDeps.isEmpty()) {
            val rawDep = needProvideDeps.pollFirst()
            val dep = wrapHelper.nonWrappedType(rawDep.typeName)
            val invokeCall = provideTypeInvokeCall(
                provideTypeCodes,
                dep,
                rawDep.qualifierAnns,
                rawDep.methodName,
                wrapHelper.isList(rawDep.typeName)
            )
            if (invokeCall == null) {
                if (provideDep == rawDep) return null
                throw ObjectNotProvidedException("Error provide type ${rawDep.typeName}")
            }

            val isBindInstanceInvoke = invokeCall.flags.provideBindInstance
            val newDeps = invokeCall.argDeps().filter {
                if (isBindInstanceInvoke && rawDep.typeName == it.typeName) {
                    // bind instance case. Argument and return type are equals
                    return@filter false
                }
                // qualifies not need to provide
                val argNonWrapped = wrapHelper.nonWrappedType(it.typeName)
                argNonWrapped is ClassName && !identifierTypes.any { it.toTypeName() == argNonWrapped }
            }
            needProvideDeps.addAll(newDeps)

            needProvideDeps = LinkedList(needProvideDeps.removeDoubles { a, b -> a == b })

            val recursiveDetected = !newDeps.isEmpty()
                    && needProvideDepsRecursiveDetector.next(needProvideDeps.hashCode())
            if (recursiveDetected) {
                throw RecursiveProviding("Error provide type ${provideDep.typeName}. Recursive providing detected.")
            }

            if (loopCount++ > MAX_PROVIDE_RESOLVE_COUNT) {
                throw StoneException(
                    "Error provide type ${provideDep.typeName}. " +
                            "Long providing loop for type. Stone library Error. "
                )
            }

            provideTypeInvokes.add(invokeCall)
        }

        if (provideTypeInvokes.filter { it.resultType() == provideDep.typeName }
                .groupBy { it.qualifierAnnotations }.size > 1) {
            throw StoneException("internal arch error")
        }
        provideTypeInvokes = LinkedList(
            provideTypeInvokes.removeDoubles { it1, it2 ->
                it1.resultType() == it2.resultType()
                        && it1.qualifierAnnotations == it2.qualifierAnnotations
            }
        )
        provideTypeInvokes.reverse()
        return provideTypeInvokes
    }

    /**
     * Generate cache control method invoke. Clean refs, change ref type and other
     *
     * @param provideMethodName predefined method name
     * @param typeName          the name of the type whose cache needs to be changed
     * @return cache control invoke call
     */
    fun invokeControlCacheForType(
        provideMethodName: String?,
        typeName: TypeName,
        qualifierAnns: Set<QualifierAnn>
    ): InvokeCall? = provideTypeInvokeCall(
        cacheControlTypeCodes,
        typeName,
        qualifierAnns,
        provideMethodName?.cacheControlMethodName,
        false
    )


    private fun provideTypeInvokeCall(
        provideTypeCodes: Map<TypeName, Set<InvokeCall>>,
        typeName: TypeName,
        qualifierAnns: Set<QualifierAnn>,
        provideMethodName: String?,
        listVariants: Boolean,
    ): InvokeCall? {
        val invokeCalls = provideTypeCodes.getOrDefault(typeName, null)
        if (invokeCalls == null || invokeCalls.isEmpty()) return null

        var filtered = invokeCalls.toList()
        if (qualifierAnns.none { it.typeName == IgnoreQualifier::class.asClassName() }) {
            filtered = invokeCalls.filter { it.qualifierAnnotations == qualifierAnns }
        }

        filtered = if (provideMethodName != null) {
            filtered.filter { provideMethodName == it.bestSequence().last().methodName }
        } else filtered

        if (!listVariants && filtered.size > 1) {
            throw IncorrectSignatureException(
                "Error provide type $typeName " +
                        ": is bound multi times.\n " +
                        filtered.joinToString(" and ")
            )
        }
        return if (!filtered.isEmpty()) {
            InvokeCall.fromVariants(
                wrapHelper,
                variants = filtered.toList(),
                qualifierAnnotations = qualifierAnns,
            )
        } else {
            null
        }
    }

    companion object {
        const val MAX_PROVIDE_RESOLVE_COUNT: Int = 10000
    }

}
